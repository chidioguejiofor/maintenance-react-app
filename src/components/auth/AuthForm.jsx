import { Form, Container, Header, Icon, Message } from 'semantic-ui-react';
import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import TopNav from '../Header';

const AuthForm = ({
  formType, handleSubmit, success, messageTitle,
  loading, handleChange, inputs, errorMessages,
  links
}) => (
  <div id="auth-form">
    <TopNav />
    <Container>
      <div className="form">
        <Form
          loading={loading}
          onSubmit={handleSubmit}
          error={!!errorMessages.length}
          success={success}
        >

          <Header as="h2" icon textAlign="center">
            <Icon name="user circle" />
            <Header.Content>{formType}</Header.Content>
          </Header>
          <Message
            header={messageTitle}
            error={!!errorMessages.length}
            success={success}
            list={errorMessages}
            hidden={!success && !errorMessages.length}
          />
          {inputs.map((input, index) => (
            <Form.Input
              key={index}
              label={input.placeholder}
              placeholder={input.placeholder}
              name={input.name}
              value={input.value}
              onChange={handleChange}
              type={input.type}
            />
          ))}
          <Form.Button
            disabled={success}
            positive
            content="Submit"
          />
          {links.map((linkObj, index) => (
            <span key={index} className="link">
              <Link to={linkObj.to}>{linkObj.caption}</Link>
            </span>

          ))}
        </Form>
      </div>

    </Container>
  </div>


);

AuthForm.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
  handleChange: PropTypes.func.isRequired,
  inputs: PropTypes.arrayOf(PropTypes.objectOf(PropTypes.string)).isRequired,
  formType: PropTypes.string,
  errorMessages: PropTypes.arrayOf(PropTypes.string),
  success: PropTypes.bool,
  messageTitle: PropTypes.string,
  loading: PropTypes.bool,
  links: PropTypes.arrayOf(PropTypes.object),
};
AuthForm.defaultProps = {
  formType: 'Sign Up',
  errorMessages: [],
  success: true,
  messageTitle: '',
  loading: false,
  links: [],
};
export default AuthForm;
