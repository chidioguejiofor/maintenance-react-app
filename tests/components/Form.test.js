import React from 'react';
import { shallow } from 'enzyme';
import { Form } from 'semantic-ui-react';
import { Link } from 'react-router-dom';
import chai from 'chai';
import AuthForm from '../../src/components/Form';

const getMounted = (options = {
  inputs: [],
  handleSubmit: () => {},
  handleChange: () => {}
}) => shallow(<AuthForm {...options} />);
describe('Testing AuthForm Component', () => {
  describe('the components when inputs property are empty rendered in the DOM', () => {
    it('should match snapshot', () => {
      const mountedObj = getMounted();
      expect(mountedObj).toMatchSnapshot();
    });

    it('should render a Form', () => {
      const mountedObj = getMounted();
      chai.expect(mountedObj.find(Form).length)
        .greaterThan(0);
    });
    it('should render an Links when links are specified ', () => {
      const options = {
        inputs: [],
        handleSubmit: () => {},
        handleChange: () => {},
        links: [
          { to: '/login', caption: 'Login' },
          { to: '/great', caption: 'Grate' },
          { to: '/love-board', caption: 'Love Board' }
        ]
      };
      const mountedObj = getMounted(options);
      chai.expect(mountedObj.find(Link).length)
        .to.equal(options.links.length);
    });
  });

  describe('the components when inputs property are empty rendered in the DOM', () => {
    it('should render a Form', () => {
      const mountedObj = getMounted();
      chai.expect(mountedObj.find(Form).length)
        .greaterThan(0);
    });

    it('should use the input props to render inputs', () => {
      const options = {
        handleChange: jest.fn(),
        error: true,
        inputs: [
          {
            placeholder: 'test',
            name: 'name',
            value: 'test 100'
          },
          {
            placeholder: 'enter test 2',
            name: 'good-name',
            value: 'final 100'
          }
        ]
      };


      const mountedObj = getMounted(options);
      chai.expect(mountedObj.find(Form.Input).length)
        .to.equal(options.inputs.length);
      mountedObj.find(Form.Input)
        .first()
        .simulate('change');
      expect(options.handleChange)
        .toHaveBeenCalled();
    });
  });
});
