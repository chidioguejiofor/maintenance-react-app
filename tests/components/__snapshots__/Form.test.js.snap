// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing AuthForm Component the components when inputs property are empty rendered in the DOM should match snapshot 1`] = `
<div
  id="auth-form"
>
  <Container>
    <div
      className="auth"
    >
      <Form
        as="form"
        error={false}
        loading={false}
        onSubmit={[Function]}
        success={true}
      >
        <Header
          as="h2"
          icon={true}
          textAlign="center"
        >
          <Icon
            as="i"
            name="user circle"
          />
          <HeaderContent>
            Sign Up
          </HeaderContent>
        </Header>
        <Message
          className="text-centered"
          error={false}
          header=""
          hidden={false}
          list={Array []}
          success={true}
        />
        <FormButton
          as={[Function]}
          content="Submit"
          control={[Function]}
          disabled={true}
          loading={false}
          positive={true}
        />
      </Form>
    </div>
  </Container>
</div>
`;
